pubsub/spec/sample_spec.rb

```diff
@@ -0,0 +1,146 @@
+  end
+
+  before(:each) do
+    topic = @pubsub.topic TOPIC_NAME
+    topic.delete if topic
+    sub = @pubsub.subscription SUB_NAME
+    sub.delete if sub
+
+    allow(Gcloud).to receive(:new).with("my-gcp-project-id").and_return(@gcloud)
+  end
+
+  it "creates topic" do
+    expect(@pubsub.topic(TOPIC_NAME)).to be nil
+
+    expect { create_topic }.to output(/#{TOPIC_NAME}/).to_stdout
+
+    topic = @pubsub.topic TOPIC_NAME
+    expect(topic.nil?).to eq(false)
+    expect(topic.exists?).to eq(true)
+    expect(topic.name).to include(TOPIC_NAME)
+  end
+
+  it "creates sub" do
+    expect(@pubsub.subscription(SUB_NAME)).to be nil
+    @pubsub.create_topic TOPIC_NAME
+
+    expect { create_subscription }.to output(/#{SUB_NAME}/).to_stdout
+
+    sub = @pubsub.subscription SUB_NAME
+    expect(sub.nil?).to eq(false)
+    expect(sub.exists?).to eq(true)
+    expect(sub.name).to include(SUB_NAME)
+    expect(sub.topic.name).to include(TOPIC_NAME)
+  end
+
+  it "creates push sub" do
+    sub_name = "my-subscription-push"
+
+    sub = @pubsub.subscription sub_name
+    sub.delete if sub
+
+    @pubsub.create_topic TOPIC_NAME
+
+    expect_any_instance_of(Gcloud::Pubsub::Topic).to \
+      receive(:subscribe).with(
+        sub_name,
+        endpoint: "https://my-gcp-project-id.appspot.com/push"
+      ).and_return(
+        @pubsub.topic(TOPIC_NAME).subscribe(
+          sub_name,
+          endpoint: "https://#{ENV['GCLOUD_PROJECT']}.appspot.com/push"
+        ))
+
+    expect { create_push_subscription }.to output(/#{sub_name}/).to_stdout
+
+    sub = @pubsub.subscription sub_name
+    expect(sub.nil?).to eq(false)
+    expect(sub.exists?).to eq(true)
+    expect(sub.name).to include(sub_name)
+    expect(sub.topic.name).to include(TOPIC_NAME)
+    sub.delete
+  end
+
+  it "publishes a message" do
+    sub = @pubsub.subscription SUB_NAME
+    @pubsub.create_subscription(
+      TOPIC_NAME,
+      SUB_NAME,
+      autocreate: true
+    ) if sub.nil?
+
+    publish_message
+
+    # Just expect no exception
+  end
+
+  it "pulls a message" do
+    message = "Test Message"
+
+    @pubsub.create_subscription(
+      TOPIC_NAME,
+      SUB_NAME,
+      autocreate: true
+    )
+    @pubsub.topic(TOPIC_NAME).publish message
+
+    $stdout = StringIO.new
+    found = false
+    (0..5).each do
+      print_messages
+      if $stdout.string.include? message
+        found = true
+        break
+      end
+    end
+    $stdout = STDOUT
+    expect(found).to eq(true)
+  end
+
+  it "prints topics" do
+    @pubsub.topic TOPIC_NAME, autocreate: true
+
+    expect { print_topics }.to output(/#{TOPIC_NAME}/).to_stdout
+  end
+
+  it "prints subs" do
+    @pubsub.create_subscription(
+      TOPIC_NAME,
+      SUB_NAME,
+      autocreate: true
+    )
+
+    expect { print_subscriptions }.to output(/#{SUB_NAME}/).to_stdout
+  end
+
+  after(:all) do
+    topic = @pubsub.topic TOPIC_NAME
+    subs = topic.subscriptions
+    subs.each(&:delete)
```

Comment:
It would be nice to be able to run multiple test simultaneously without them stepping on each others toes. In fact, I had to change the travis config for this repo to run test serially, otherwise it was running the pub/sub test for both versions of ruby at the same time. I'd need to update the test to add a random element to the topic name to get past that here.